source/h2_decoder.c

```diff
@@ -15,48 +15,622 @@
+
+        uint8_t type; /* aws_h2_frame_type */
+        uint8_t flags;
+        uint8_t padding_len;
+    } frame_in_progress;
 
     /* User callbacks and settings. */
-    struct aws_http_decoder_vtable vtable;
-    void *user_data;
+    struct aws_h2_decoder_vtable vtable;
+    void *userdata;
 };
 
+/***********************************************************************************************************************
+ * Public API
+ **********************************************************************************************************************/
+
+static void s_decoder_reset_state(struct aws_h2_decoder *decoder);
+
 struct aws_h2_decoder *aws_h2_decoder_new(struct aws_h2_decoder_params *params) {
     AWS_PRECONDITION(params);
     (void)params;
 
-    struct aws_h2_decoder *decoder = aws_mem_calloc(params->alloc, 1, sizeof(struct aws_h2_decoder));
-    if (!decoder) {
-        return NULL;
+    struct aws_h2_decoder *decoder = NULL;
+    void *scratch_buf = NULL;
+
+    void *allocation = aws_mem_acquire_many(
+        params->alloc, 2, &decoder, sizeof(struct aws_h2_decoder), &scratch_buf, s_scratch_space_size);
+    if (!allocation) {
+        goto failed_alloc;
     }
 
     decoder->alloc = params->alloc;
-    decoder->user_data = params->user_data;
+    decoder->logging_id = NULL;
+
     decoder->vtable = params->vtable;
+    decoder->userdata = params->userdata;
+
+    decoder->scratch = aws_byte_buf_from_array(scratch_buf, s_scratch_space_size);
+
+    decoder->hpack = aws_hpack_context_new(params->alloc, s_hpack_dynamic_table_initial_size);
+    if (!decoder->hpack) {
+        goto failed_new_hpack;
+    }
+
+    AWS_ZERO_STRUCT(decoder->state);
+    decoder->state.name = "<NONE>";
+    s_decoder_reset_state(decoder);
 
     return decoder;
+
+failed_new_hpack:
+    aws_mem_release(params->alloc, allocation);
+failed_alloc:
+    return NULL;
 }
 
 void aws_h2_decoder_destroy(struct aws_h2_decoder *decoder) {
+    aws_hpack_context_destroy(decoder->hpack);
     aws_mem_release(decoder->alloc, decoder);
 }
 
 int aws_h2_decode(struct aws_h2_decoder *decoder, struct aws_byte_cursor *data) {
     AWS_PRECONDITION(decoder);
     AWS_PRECONDITION(data);
-    (void)decoder;
-    (void)data;
 
-    return AWS_OP_ERR;
+    while (data->len) {
+        const uint64_t bytes_required = decoder->state.bytes_required == WAIT_FOR_FULL_PAYLOAD
+                                            ? decoder->frame_in_progress.payload_len
+                                            : decoder->state.bytes_required;
+        int err = AWS_OP_SUCCESS;
+        if (!decoder->scratch.len && data->len >= bytes_required) {
+            /* Easy case, there is no scratch and we have enough data, so just send it to the state */
+
+#if DEBUG_BUILD
+            const char *current_state = decoder->state.name;
+            const uint64_t pre_state_data_len = data->len;
+#endif
+
+            err = decoder->state.fn(decoder, data);
+
+#if DEBUG_BUILD
```

Comment:
I'd love to be able to get these logging statements on a production system without having to recompile the application.